File: Data Structures & Algorithms/split-array-largest-sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number}
     */
    splitArray(nums, k) {
        let length=nums.length;
        let min_sum=Math.max(...nums);
        let max_sum=nums.reduce((acc,curr)=>acc+curr,0);
        const check = (limit)=>{
            let cnt=1, sum=0;
            for(let i=0;i<length;i++){
                sum+=nums[i];
                if(sum>limit){
                    cnt++;
                    sum=nums[i];
                }
            }
            return cnt<=k;
        }
        //linear search --------->
        // for(let sum=min_sum; sum<=max_sum; sum++){
        //     if(check(sum)) return sum;
        // }
        // return -1;

        //binary search------>
        let left=min_sum, right=max_sum;
        while(left<=right){
            let mid = Math.floor((left+right)/2);
            if(!check(mid)) left = mid+1;
            else right=mid-1;
        }
        return left;
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-1.java
class Solution {
    public int maxSubArray(int[] nums) {
        int n=nums.length;
        if(n==1) return nums[0];
        int maxsum=nums[0],currsum=nums[0];
        for(int i=1;i<n;i++){
            currsum=currsum+nums[i];
            currsum=Math.max(currsum,nums[i]);
            maxsum=Math.max(maxsum,currsum);
        }
        return maxsum;
    }
}
